Backfill the assets removed from minified WordPress bundles

bgrgicak opened the pull request:
## Motivation for the change, related issues

To achieve [offline support](https://github.com/WordPress/wordpress-playground/pull/1483) we need to download all WordPress files into the Playground filesystem. Today Playground downloads WordPress partially on boot and downloads assets as they are needed using fetch requests.

This PR downloads all assets into the Playground filesystem without blocking the boot process. 

## Implementation details

Zip files are generated using the WordPress build script and are accessible in the `/wp-VERSION/` folder like any other assets.

The download is triggered after WordPress is installed. The fetch is async and won't block the boot process. 

In case some of the assets are needed immediately on boot they will still be downloaded using fetch to ensure they are available to WordPress. 

Once the assets are downloaded they are stored in the Playground filesystem and will be served from there.

## Testing Instructions (or ideally a Blueprint)

- Run `npx nx bundle-wordpress:nightly playground-wordpress-builds`
- Confirm that a `packages/playground/wordpress-builds/public/wp-nightly/wordpress-static.zip` file was created
- [Load Playground using this blueprint](http://127.0.0.1:5400/website-server/?php=8.0&wp=nightly)
- In network tools see that the `wordpress-static.zip` file was downloaded
- Open `/wp-admin/` 
- In network tools confirm that static assets like _thickbox.js_ were loaded from the service worker instead of a fetch

bgrgicak commented:
@adamziel I addressed the SQLite issue by not including it in the zip. Also we won't override existing files anymore to prevent this from happening to other files.

[You asked me to better document the backfill code](https://github.com/WordPress/wordpress-playground/pull/1532#discussion_r1672101261), I did it here https://github.com/WordPress/wordpress-playground/pull/1604/files#diff-5466414cb25ca570e2095b662e08efbdcb5edd0aee22de36ee9c2c2eb2cdfb50R198-R225

adamziel commented on packages/playground/common/src/index.ts:
```diff
@@ -21,7 +21,8 @@ const tmpPath = '/tmp/file.zip';
 	zipPath: string | File,
-	extractToPath: string
+	extractToPath: string,
+	overwrite = true
```
Let's rename this to:

```suggestion
	overwriteFiles = true
```

Otherwise it's unclear what happens when a directory name conflicts – e.g. is the old directory wiped out?

adamziel commented on packages/playground/common/src/index.ts:
```diff
@@ -45,14 +47,23 @@ export const unzipFile = async (
+				for ($i = 0; $i < $zip->numFiles; $i++) {
+					$filename = $zip->getNameIndex($i);
+					$fileinfo = pathinfo($filename);
+					$extractFilePath = $extractTo . DIRECTORY_SEPARATOR . $filename;
```
Let's trim any trailing `$extractTo` doesn't end with `DIRECTORY_SEPARATOR`.

adamziel replied:
Also, I'm not sure if `DIRECTORY_SEPARATOR` is any useful these days. Windows seems to understand forward slashes just fine, and in WASM we're assuming linux anyway so `DIRECTORY_SEPARATOR` is always `/`.

adamziel commented on packages/playground/remote/src/lib/boot-playground-remote.ts:
```diff
@@ -232,6 +239,25 @@ export async function bootPlaygroundRemote() {
 	}
 
+	/**
+	 * If the browser is online we can download WordPress assets asynchronously to speed up the boot process.
```
I got a bit confused why that's optional – this should clarify it:

```suggestion
	 * If the browser is online we download the WordPress assets asynchronously to speed up the boot process.
```

Also, let's clarify this comment talks about the minified WordPress bundles and that we're not doing either of these when a full production build is used (since it ships all the static assets already).

adamziel replied:
I'm reopening this conversation because this isn't addressed yet:

> Also, let's clarify this comment talks about the minified WordPress bundles and that we're not doing either of these when a full production build is used (since it ships all the static assets already).

bgrgicak replied:
Done [2ef0bcd](https://github.com/WordPress/wordpress-playground/pull/1604/commits/2ef0bcdf0ac36f63927b18025ef96239b12509ce). I kept that part short because it's described in detail in the function docblock.

adamziel commented on packages/playground/remote/src/lib/boot-playground-remote.ts:
```diff
@@ -232,6 +239,25 @@ export async function bootPlaygroundRemote() {
+	 * If the browser is online we can download WordPress assets asynchronously to speed up the boot process.
+	 * Missing assets will be fetched on demand from the Playground server until they are downloaded.
+	 *
+	 * If the browser is offline, we download WordPress assets synchronously to ensure the Playground is fully functional on load.
```
How are we downloading assets if the browser is offline?

Also, nitpick: Synchronous means blocking, meaning the event loop would be stopped until the download is complete. In here, however, we're not pausing the entire worker but just awaiting the download before continuing the boot process.

bgrgicak replied:
The comment was misleading, this is what it needs to say: _ If the browser is offline, we backfill WordPress assets synchronously from the cache to ensure Playground is fully functional before boot finishes._

adamziel replied:
I'm reopening this conversation because this isn't addressed yet:

> Synchronous means blocking, meaning the event loop would be stopped until the download is complete. In here, however, we're not pausing the entire worker but just awaiting the download before continuing the boot process.

bgrgicak replied:
I get it now. In my mind, the comment sounded the same, but await and synchronous definitely aren't the same. [3e4cd07](https://github.com/WordPress/wordpress-playground/pull/1604/commits/3e4cd07ccc82009406b85f8888a0c9dc7e1b5059)

adamziel commented on packages/playground/remote/src/lib/worker-thread.ts:
```diff
@@ -186,6 +187,86 @@ export class PlaygroundWorkerEndpoint extends PHPWorker {
+ *
+ * To load Playground faster, we ship minified WordPress builds without most CSS files, JS files, and other static assets.
+ *
+ * Instead, the build contains a list of remote assets in the wordpress-remote-asset-paths file which is located in the WordPress document root.
```
Nitpick: Let's wrap these comment lines at 80 or 120 characters

adamziel commented on packages/playground/remote/src/lib/worker-thread.ts:
```diff
@@ -186,6 +187,86 @@ export class PlaygroundWorkerEndpoint extends PHPWorker {
+ *
+ * ## Background
+ *
+ * To load Playground faster, we ship minified WordPress builds without most CSS files, JS files, and other static assets.
```
Let's clarify that minified builds are not mandatory but only used by default:

```suggestion
 * To load Playground faster, we default to minified WordPress builds shipped without most CSS files, JS files, and other static assets.
```

adamziel commented on packages/playground/remote/src/lib/worker-thread.ts:
```diff
@@ -186,6 +187,86 @@ export class PlaygroundWorkerEndpoint extends PHPWorker {
+}
+
+/**
+ * Downloads static assets that are removed from minified WordPress builds.
```
Let's explain the gist of it in the first paragraph:

```suggestion
 * Downloads and unzips a ZIP bundle of all the static assets removed from
 * the currently loaded minified WordPress build. Doesn't do anything if the
 * assets are already downloaded or if a non-minified WordPress build is loaded.
```

adamziel commented on packages/playground/remote/src/lib/worker-thread.ts:
```diff
@@ -186,6 +187,86 @@ export class PlaygroundWorkerEndpoint extends PHPWorker {
+ * Instead, the build contains a list of remote assets in the wordpress-remote-asset-paths file which is located in the WordPress document root.
+ * Playground uses this list to determine if it should fetch the assets on demand.
+ *
+ * When Playground requests a static asset that is not in the minified build, we download the asset from the Playground.WordPress.net server.
```
This comment gives all the right information, but I find the structure confusing. I feel it would be more useful if it made the steps crystal clear and explained the exact timeline and how this function fits in:

```
Here's how Playground boot flow and the assets backfilling process fits together:

1. WordPress minified build is loaded
2. The WordPress site is first rendered in the browser
3. The missing static assets are backfilled by this function

## WordPress minified build is loaded

This entire function is only relevant for minified builds where most CSS, JS, and other static assets are missing after the initial load. Each minified build ships a `wordpress-remote-asset-paths` file that ...

See: (the dockerfile lines producing the minified build)

## The WordPress site is first rendered in the browser

Many CSS and JS files are missing at this point. When the browser tries to load a CSS file, the request goes to the service worker which then decides how to handle it based on the `/wordpress/wordpress-remote-asset-paths` file. (...).

See (the function or PR where `wordpress-remote-asset-paths` is documented)

## The missing static assets are backfilled by this function

Right after the first page load, we request a ZIP-ped bundle of all the static assets missing from the currently loaded minified build. 

This means playground.wordpress.net ships two copies of the static assets:

* Unzipped files that can be directly fetched, e.g. `/wp-6.2/wp-content/themes/twentytwentyone/style.zip`
* Zipped bundle used for backfilling, e.g. `/wp-6.2/wordpress-static.zip`

The goal is to unzip them in the WordPress filesystem and avoid network requests every time a static file is needed. Furthermore, we need those files available locally to run Playground in offline mode.

This function uses the same `wordpress-remote-asset-paths` file as we've discussed above to (...)
```

I know this may seem like a lot, but this function is an important part of the flow. It's highly useful to provide  comprehensive documentation, and it's totally fine if the comment takes more lines and effort than the code itself. I sometimes start with the comment and only then write the actual implementation – the process of explaining it first reveals gaps in my thinking.

bgrgicak replied:
Let me know if this is better https://github.com/WordPress/wordpress-playground/pull/1604/commits/1600eadac2ca720877fc6ac8c7823fc0456f0019#diff-5466414cb25ca570e2095b662e08efbdcb5edd0aee22de36ee9c2c2eb2cdfb50

adamziel replied:
It's better! I added one more suggestion below – LMK what you think

bgrgicak replied:
I see now. I write it as dry technical documentation, while your style is more like a blog post.
The example also helps with explaining how it works.

adamziel commented:
Most of my notes are about comments and code style. The only technical blocker I see is this:

> If the browser is offline, we download WordPress assets synchronously to ensure the Playground is fully functional on load.

How would it download the static assets when the browser is offline? Would it use the `zip` file cached in the service worker, reject the promise, and display a reasonable error message if that file not cached yet? If yes, that sounds fine.

bgrgicak commented on packages/playground/remote/src/lib/boot-playground-remote.ts:
```diff
@@ -232,6 +239,25 @@ export async function bootPlaygroundRemote() {
+	 * If the browser is online we can download WordPress assets asynchronously to speed up the boot process.
+	 * Missing assets will be fetched on demand from the Playground server until they are downloaded.
+	 *
+	 * If the browser is offline, we download WordPress assets synchronously to ensure the Playground is fully functional on load.
```
```suggestion
	 * If the browser is offline, we backfill WordPress assets synchronously from the cache to ensure Playground is fully functional before boot finishes.
```

bgrgicak commented:
> How would it download the static assets when the browser is offline? Would it use the zip file cached in the service worker, reject the promise, and display a reasonable error message if that file not cached yet? If yes, that sounds fine.

The comment was misleading. The difference is that in offline mode boot waits for the backfill to finish, and when offline it doesn't. 

For offline support to work, the `wordpress-static.zip` needs to be cached, so we expect it to be available.

adamziel commented on packages/playground/remote/src/lib/boot-playground-remote.ts:
```diff
@@ -232,6 +239,26 @@ export async function bootPlaygroundRemote() {
+			webApi.backfillStaticFilesRemovedFromMinifiedBuild();
+		});
+	} else {
+		webApi.setProgress({
```
```suggestion
		// Note this will run even if the static files are already in place, e.g. when running
		// a non-minified build or an offline site. It doesn't seem like a big problem worth introducing
		// a new API method like `webApi.needsBackfillingStaticFilesRemovedFromMinifiedBuild().
		webApi.setProgress({
```

adamziel commented on packages/playground/remote/src/lib/worker-thread.ts:
```diff
@@ -186,6 +187,104 @@ export class PlaygroundWorkerEndpoint extends PHPWorker {
+ * the currently loaded minified WordPress build. Doesn't do anything if the
+ * assets are already downloaded or if a non-minified WordPress build is loaded.
+ *
+ * # Asset Loading
+ *
+ * ## WordPress minified build is loaded
+ *
+ * To load Playground faster, we default to minified WordPress builds shipped
+ * without most CSS files, JS files, and other static assets.
+ *
+ * ## Static assets are fetched on demand before backfilling
+ *
+ * The build contains a list of remote assets in the wordpress-remote-asset-paths file
+ * which is located in the WordPress document root.
+ * Playground uses this list to determine if it should fetch the assets on demand.
+ *
+ * When Playground requests a static asset that is not in the minified build,
+ * we download the asset from the Playground.WordPress.net server.
+ *
+ * ## Backfilling on boot
+ *
+ * At the end of the Playground boot process (see `bootPlaygroundRemote`),
+ * Playground starts the backfill process by calling `backfillStaticFilesRemovedFromMinifiedBuild`.
+ *
+ * ### Prevent backfilling if assets are already available
+ *
+ * The function checks if the wordpress-remote-asset-paths file exists and is not empty.
+ * The wordpress-remote-asset-paths file is emptied after the assets are downloaded,
+ * so we can assume that the assets are already downloaded if the file is empty.
+ * If the wordpress-remote-asset-paths file doesn't exist,
+ * this means that the WordPress files didn't originate from a minified Playground build.
+ * We backfill only for minified builds, so we skip the backfill process in this case.
+ *
+ * ### Downloading assets during backfill
+ *
+ * Each WordPress release has a corresponding static assets directory on the Playground.WordPress.net server.
+ * The file is downloaded from the server and unzipped into the WordPress document root.
+ *
+ * ### Skipping existing files during unzipping
```
This is a nice comment. I gave it a read and still thought some parts needed more context so I iterated on it like this:

```suggestion
 * ## Asset Loading
 *
 * To load Playground faster, we default to minified WordPress builds shipped
 * without most CSS files, JS files, and other static assets.
 *
 * When Playground requests a static asset that is not in the minified build, the service
 * worker consults the list of the assets removed during the minification process. Such
 * a list is shipped with every minified build in a file called `wordpress-remote-asset-paths`.
 *
 * For example, when `/wp-includes/css/dist/block-library/common.min.css` isn't found
 * in the Playground filesystem, the service worker looks for it in `/wordpress/wordpress-remote-asset-paths`
 * and finds it there. This means it's available on the remote server, so the service
 * worker fetches it from an URL like:
 *
 * https://playground.wordpress.net/wp-6.5/wp-includes/css/dist/block-library/common.min.css
 *
 * ## Assets backfilling
 *
 * Running Playground offline isn't possible without shipping all the static assets into the browser.
 * Downloading every CSS and JS file one request at a time would be slow to run and tedious to maintain.
 * This is where this function comes in!
 *
 * It downloads a zip archive containing all the static files removed from the currently running
 * minified build, and unzips them in the Playground filesystem. Once it finishes, the WordPress
 * installation running in the browser is complete and the service worker will no longer have
 * to backfill any static assets again.
 *
 * This process is started after the Playground boots (see `bootPlaygroundRemote`) and the first
 * page is rendered. This way we're not delaying the initial Playground paint with a large download.
 *
 * ## Prevent backfilling if assets are already available
 *
 * Running this function twice, or running it on a non-minified build will have no effect.
 *
 * The backfilling only runs when a non-empty `wordpress-remote-asset-paths` file
 * exists. When one is missing, we're not running a minified build. When one is empty,
 * it means the backfilling process was already done – this function empties the file
 * after the backfilling is done.
 *
 * ### Downloading assets during backfill
 *
 * Each WordPress release has a corresponding static assets directory on the Playground.WordPress.net server.
 * The file is downloaded from the server and unzipped into the WordPress document root.
 *
 * ### Skipping existing files during unzipping
```

adamziel reviewed (approved):
The implementation seems great so I'm approving. Thank you @bgrgicak! Let's still address the notes I left about the documentation and inline comments before merging.

bgrgicak commented:
@adamziel I addressed your comments and will merge it now to unblock development.
Let me know if the comments require more changes and I can push them to another PR.
